turbopack/crates/turbo-tasks-fs/src/lib.rs

```diff
@@ -401,7 +370,72 @@ impl FileSystem for DiskFileSystem {
             .collect::<Result<_>>()
             .with_context(|| format!("reading directory item in {}", full_path.display()))?;
 
-        Ok(DirectoryContent::new(entries))
+        Ok(InternalDirectoryContent::new(entries))
+    }
+}
+
+impl Debug for DiskFileSystem {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "name: {}, root: {}", self.name, self.root)
+    }
+}
+
+#[turbo_tasks::value_impl]
+impl FileSystem for DiskFileSystem {
+    #[turbo_tasks::function(fs)]
+    async fn read(&self, fs_path: Vc<FileSystemPath>) -> Result<Vc<FileContent>> {
+        let full_path = self.to_sys_path(fs_path).await?;
+        self.register_invalidator(&full_path)?;
+
+        let _lock = self.lock_path(&full_path).await;
+        let content = match retry_future(|| File::from_path(full_path.clone()))
+            .instrument(tracing::info_span!(
+                "read file",
+                path = display(full_path.display())
+            ))
+            .await
+        {
+            Ok(file) => FileContent::new(file),
+            Err(e) if e.kind() == ErrorKind::NotFound || e.kind() == ErrorKind::InvalidFilename => {
+                FileContent::NotFound
+            }
+            Err(e) => {
+                bail!(anyhow!(e).context(format!("reading file {}", full_path.display())))
```

Comment:
TIL: I thought `bail!` and `anyhow!` would stringify this with `Display`/`ToString`, and you needed to call `anyhow::Error::new` to preserve the underlying error, but it turns out that there's a crazy auto-ref specialization hack they use to avoid that:

https://github.com/dtolnay/anyhow/blob/c378a2c32e0425bc981acd40c2e829a1ec62c0b0/src/kind.rs#L1-L45